src/coreclr/jit/optimizer.cpp

```diff
@@ -6029,34 +6029,14 @@ bool Compiler::optNarrowTree(GenTree* tree, var_types srct, var_types dstt, Valu
                             dstt = varTypeToSigned(dstt);
                         }
 
-                        if ((oprSize == dstSize) &&
-                            ((varTypeIsUnsigned(dstt) == varTypeIsUnsigned(oprt)) || !varTypeIsSmall(dstt)))
-                        {
-                            // Same size and there is no signedness mismatch for small types: change the CAST
-                            // into a NOP
-
-                            JITDUMP("Cast operation has no effect, bashing [%06d] GT_CAST into a GT_NOP.\n",
-                                    dspTreeID(tree));
-
-                            tree->ChangeOper(GT_NOP);
-                            tree->gtType = dstt;
-                            // Clear the GTF_UNSIGNED flag, as it may have been set on the cast node
-                            tree->gtFlags &= ~GTF_UNSIGNED;
-                            tree->AsOp()->gtOp2 = nullptr;
-                            tree->gtVNPair      = op1->gtVNPair; // Set to op1's ValueNumber
-                        }
-                        else
-                        {
-                            // oprSize is smaller or there is a signedness mismatch for small types
+                        // oprSize is smaller or there is a signedness mismatch for small types
+                        // Change the CastToType in the GT_CAST node
+                        tree->CastToType() = dstt;
```

Comment:
What IR does this create for the example case we were looking at?

Reply:
It used to be:
```
    [000002] ----G+-----                         *  CAST      long <- int
    [000001] n---G+-----                         \--*  IND       byte  
    [000000] H----+-----                            \--*  CNS_INT(h) long   0x7ffa2213f21a static Fseq[s_13]

to

    [000002] ----G+-----                         *  NOP       byte  
    [000001] n---G+-----                         \--*  IND       byte  
    [000000] H----+-----                            \--*  CNS_INT(h) long   0x7ffa2213f21a static Fseq[s_13]
```

Now it's:
```
    [000002] ----G+-----                         *  CAST      long <- int
    [000001] n---G+-----                         \--*  IND       byte  
    [000000] H----+-----                            \--*  CNS_INT(h) long   0x7ffa22c7f21a static Fseq[s_13]

to

    [000002] ----G+-----                         *  CAST      int <- byte <- int
    [000001] n---G+-----                         \--*  IND       byte  
    [000000] H----+-----                            \--*  CNS_INT(h) long   0x7ffa22c7f21a static Fseq[s_13]

```

Reply:
That's still an illegal transformation in this case since the value was an argument. The new cast does not guarantee widening into the upper 32 bits like the original cast does.

Reply:
@jakobbotsch are you suggesting to remove the whole transformation?

Reply:
It seems there are two separate questions:
1. Who is invoking `optNarrowTree` in this way that changes semantics of the program? The transformation is questionable to me but it's because the end result is a change change in behavior -- and this PR still has that problem.
2. Can we get rid of the weird passthrough `GT_NOP`? Yes, and this PR probably accomplishes that, but it seems a better way would be to refactor `optNarrowTree` to take a `GenTree**` so that it can just replace the use of the cast with the operand in this situation.

I'm more worried about (1) than (2), though it would also be nice to clean up (2).